Always use aten.constant_pad_nd for mm padding

oniononion36 opened the pull request:
Summary: From experiment, it seems like aten.constant_pad_nd has better QPS compared to torch.cat. The qps gain for ig ctr is ~10%, and ~5% for oc.

Test Plan:
```
buck2 run mode/opt -c fbcode.nvcc_arch=a100 //caffe2/torch/fb/model_transform/experimental/benchmark:mts_gpu_benchmark -- --model-path=manifold://ads_storage_fblearner/tree/user/facebook/fblearner/predictor/585279927/480/gpu_lowering/input.predictor.disagg.gpu.merge --lower-backend=AOT_INDUCTOR
```
```
buck2 run mode/opt //caffe2/torch/fb/model_transform/experimental/benchmark:mts_gpu_benchmark -- --model-path=manifold://ads_storage_fblearner/tree/user/facebook/fblearner/predictor/588102397/1500/gpu_lowering/input.predictor.disagg.gpu.merge --lower-backend=AOT_INDUCTOR
```

Differential Revision: D64271583




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov @rec

oniononion36 commented:
@pytorchbot label "topic: not user facing"

oniononion36 commented:
I think it is due to different lowering, in my gpu traces, I can see that using torch.cat for padding results in a couple more triton_poi_fused kernels and a couple other triton_poi_fused kernels also take way longer to execute.

eellison commented:
@oniononion36 could we try updating the lowering for aten.cat instead ? see, this function https://github.com/pytorch/pytorch/blob/main/torch/_inductor/lowering.py#L1518

oniononion36 commented:
After offline discussion, we agreed to first proceed with diff while the investigation into aten.cat is still ongoing.

eellison reviewed (approved):
todo: follow up with minimal repro

oniononion36 committed:
Always use aten.constant_pad_nd for mm padding (#137820)

Summary:

From experiment, it seems like aten.constant_pad_nd has better QPS compared to torch.cat. The qps gain for ig ctr is ~10%, and ~5% for oc.

This gain comes from the fact that the generated cat kernel is not optimized and runs considerably slower than that of constant_pad_nd. See D64571598 for a base repro.

Test Plan:
```
buck2 run mode/opt -c fbcode.nvcc_arch=a100 //caffe2/torch/fb/model_transform/experimental/benchmark:mts_gpu_benchmark -- --model-path=manifold://ads_storage_fblearner/tree/user/facebook/fblearner/predictor/585279927/480/gpu_lowering/input.predictor.disagg.gpu.merge --lower-backend=AOT_INDUCTOR
```
```
buck2 run mode/opt //caffe2/torch/fb/model_transform/experimental/benchmark:mts_gpu_benchmark -- --model-path=manifold://ads_storage_fblearner/tree/user/facebook/fblearner/predictor/588102397/1500/gpu_lowering/input.predictor.disagg.gpu.merge --lower-backend=AOT_INDUCTOR
```

Differential Revision: D64271583

eellison commented:
@pytorchbot merge

shunting314 commented:
> The qps gain for ig ctr is ~10%, and ~5% for oc.

Is QPS here e2e QPS or just measuring specific kernels?

oniononion36 commented:
Hi Shunting, the QPS here is e2e. In the min-repro kernel example we used, the QPS gain is ~40%

eellison commented:
Was talking with @shunting314, my suspicion is that the speedup is due to different benchmarking within pad_mm, bc constant_pad_nd is likely much more optimized for case of large tensor + small fill value compared to torch.cat. In the eager benchmarking, there is an additional aten.zeros kernel as well. 

I imagine the inductor codegen is less affected.
